### DISCUSSION TITLE
ESP32 does not connect to Wireguard through LAN8720

### DISCUSSION THREAD

**Discussion by MahdiBabaie** [espressif/arduino-esp32#7701] [Q&A]
Hello guys
My components are an ESP32-Wroom and a LAN8720. I connect my ESP32 module to LAN8720 as follow:

3V3 | Vcc
GND | GND / RBIAS
D18 | MDIO
D19 | TXD0
D21 | TXEN
D22 | TXD1
D23 | MDC
D25 | RXD0
D26 | RXD1
D27 | CRS_DV

I can connect to my access point (in my case a modem) and access the internet. But I can not connect to my Wireguard server.
Configuration of my Wireguard is correct because I can connect to my wireguard server using WiFi connection.
Is it possible to connect to Wireguard server through LAN8720 at all?

**Comment by mrengineer7777** [comment]
Where is the 50Mhz common clock?

**Reply by MahdiBabaie** [reply]
I have a LAN8720 module and it has a 50MHz crystal

**Reply by MahdiBabaie** [reply]
The LAN8720 module like this:
data:image/jpeg;base64,/9j/4AAQSkZJRgABAQAAAQABAAD/2wCEAAkGBxMSEhMSExMWFhIXGRoXFhYXFxYXGBkXGBgZGh0ZGhkYHCggGxslHhYXITIhJSkrLi4uGB8zODMsNyguLisBCgoKDg0OGhAQGi0lHR0tLS0rLS0tLS0tLS0tLS0tLS0tLS0tLS0tLS0tKy0tLS0tLS0tLS0tLSstLS0tLS0tNf/AABEIAOEA4QMBIgACEQEDEQH/xAAcAAEAAgMBAQEAAAAAAAAAAAAABQYDBAcCAQj/xAA+EAACAgECBAUBBQUHBAIDAAABAgADEQQhBRIxQQYTIlFhcTJCUpGhByNigbEUM0NywfDxFdHS4TSSJILC/8QAFwEBAQEBAAAAAAAAAAAAAAAAAAECA//EACARAQEBAQACAgIDAAAAAAAAAAABEQIxQRIhUfBhcaH/2gAMAwEAAhEDEQA/AO4xEQEREBERAREQEREBERAREQEREBERAREQEREBERAREQEREBERAREQEREBERAREQEREBERAREQEREBERAREQEREBERAREQEREBERAREQEREBERAREQEREBERAREQEREBERARKL4l/afpdLYKqwb2BxYayOVB3AY7M3wNvcia3j7xPp7+FPdTqGHNjAU8jHfBrfO46nbuQMZ6ELhpfEGlseytL62ev7YDDb5+QO5GcHaadvE7LyU0/orH272HT/ACg/1P5d5+cNJxBvMSxCVKEMoIGSR7jsOxxOgcF8cHV40+ptqqVANx+7qO/RgNubfp0+BgmOpfSzHVKuJVUIiK7XsdwFPOxyepJOAPqfzkyDKNXxWrTALpV81iRm07hvhADuD+LOPlpOaK217A99gq2LLQGA9P4n7t/Pb4EzKtieiaeh4ilxbkyVX72MKf8AKe/16TcmmSIiAiIgIiICIiAiIgIiICIiAiIgIiICIlB8Z/tKp0vNVpuW6/oTnNSHp6iPtN/CP5kdIFt47xyjR1m2+wIvYdWY+yqN2M4n40/aPfrOaqrNOmO3KD63H8bDoP4Rt7kyq8Y4tdqrDbfY1jnuegHsoGyj4EcO4S9wL/YpU4a1geUHGeVQAWd8b8ign6DeXBn8PU1M1jWI1pRP3VKZzZaThQcAnlG5IAycASZ1nh+66htTqL0VQ3l+Up/dUKuCwdl9IIwByglididiBh4twqurThqHTkZ/Tf5oD2hFJLLuAq5PL5YPODs3MSMQ3EtZbcuj05sNKjDK2Sqg55BYMdD6AuR0wRt0mbqxHcQuqc116RXbB5Tac89hxsBWNgvXbcnBzjfPy3mpY02Jy2j7uRgA9OYg7H3EuPGuIC3U1aLRhabmrSvzygHQeoKUGUXPP68FmI7bGV7WcDp0ZYauwvacgU1kqX7B7GYZqU9d/Uc9DuInRYl/DPjC/QrgCsjJ9TL6lz2rOcL36g9Z0XhGo09qee9wsJwxqVg1rEjI8wZ9H/7Yx7CcO0QsINnlM1deQ2Q2K1O4yT7bY74khw3ibJatlWOVTnO/rGc8p7hT87/SavMpLj9B6Dit9oKVolY6843WtfbfZm+TgfBkuutrorQGw2k7rg87PnuPj56TmHAvFza4jTMK6So5uQNy1kZGWbmOSdxsxx33ltp4hXQQtWLbM+qw5KjH3a8bsfnoP0mPueVzV0Bn2RGlsutsV2by69yte3M49znqN+369ZLzTJERAREQEREBERAREQEREBERASP43xqjSVm2+wIvbuzH2VRux+BKn4z/AGlU6Xmq0+LrxsTn93Wf4iPtEfhH8yJyniNWt1udVaxdj9kMcNy/wJjCrvt0zjv3LiZ8aftHu1fNVTmnTnbAP7xx/Gw6D+EfzJlGVSSABknAAHc9gAO8y6XSvZYKkUm1jy8pwDnuDnGMYOc9MSz8E4XTRqaVssP9qV1sXB5URq3zhg4HOcqdgRnGAR1l3Bj4T4Y9S+evM5JAp5ggBA5itj5+3j/Bry+45jWN5OanWis3eRXXqLBWVFPMeaipvtLXWvKCg3J5VWwH7YOzCJ1vAbg3l0WNWrepqXfmCKrc3m1OoPmVqTnmQeYnN6huScOj1un4ccUrXqtUfsW42pYry8qrhg5yWIZCM4XfBIhG9wehKfPfWPWK3CltMoIrbmBKsrIpXKhWUNUSytgE5yDr8RqTVrWNJWPJFiLYSmL0rr5RkgNyWKAeYsm48z1BV3O7XwhbUXUa5UrdTzOMvWrh8AecFGKrCWRsqQWByyjDOvzW6rU2W/2bToNHUjAB3KIAHZmqIZRjLHPKU5mO/qYcxIU/hmq8jVvfWecpW1ivyjCt2Zeuwz177dJI+EVqOn1mq1rJZWeflDAW3K+QW5eb+7Z8j1Z6A43OZbNfp9FXaUJrs17jyrPu1u4O5Y1qRUWOQfSc/fUbtKLxXhtnJetqeQWtdygxsqpzDAGxU8y4YbEAkdgc2Kz6rQarX1VtUEGkYkVUoxUqV3Z7QeuCCC5zk55Rk7wnEq6qCiUXGy0ZW230+SB+Fe/XB5/yPtOcU47c2k0ukq5UNpVmVVXDEAKOb8QPLkqwPXvnaa4nw/Sae2hNPRXZq7a0NOnDegOPvOWODaSSQMhRy4G43m4qmab0AcyspJ+8pHN/ECRv+uAwnVv2ba2wrZUE824EGtmzy1ocjBB64K5A26n2E5T4i117ahV1T81iHlKjlK1AHJxyek5Ayeu2J3/9lXBHo072uVPnlXXlYNhANssNiSWY7Z7TfmJ4WHhXB/LbzbGNl56ue3wvsPpJaIkQiIgIiICIiAiIgJi1GoWtSzsFUdz/AL6yN4nx1K8qmHcdd/Sv1Pc/A/SV572sPPY3M3YHoP8AKvb+szesanOpLX8aezIqyifiP2z9B90fr9JFcO1dlD8ynKk+pSdm+fhvmCzdpg5vcic/lW5zFp1PifTVUm62wVqOqt9rPsFG7E4OMZ6Tkni39oep1zHT6UNXSdsL/eWD+JgfSvwD9SRLhxDRJYhSxA6Hsf8AQ9pUV0H9iVjWOU5LJdjORj+6vz0rPTnBAXYsBjmO/lc+vLPx+0Zwrw8qYa3DONwPuL/5H67fHeT2f16medFq69TkKPL1CDNlD4BHyvYr3B9sHpjOjxPi6VELvZaThKk9RJ+ijc7e38jvOc7+X9/j3+/438ceeK8JV/34c1PX6hcAcgruCeXc4x9odP0mDQeIX5v7PxAcuT/emuskBgADsNugIsTJx2YYxG28eet2ezksuxhKwSa6NxuxRsWP25RlR+I45RA6zW2XObLXZ3PVmO/09gPgbTtxL7c+rPTpzaZqlqrYNbpMl/OCLz1ltw4srIZeUnAuU/pgzB/ZBWw1lZUP6h55TlqsGdxqBWP3LHB//IrIUnPMF5sml+HvED6Yhcc9JOXrJK7Njm5WB9OQACDlTgZBwMXjhzVFF1WncLQHPPWvnH1FRk8v+DuyZVSR3BM0y1eE8Ge+wPrFN165FWjXHIi9ncL6VQ4JxkBlC45sqs3tFraq0/siG17M1JlGaoaRSeRSWsTKNkJs+7vvyqSM49BrEvpbygasMyNUtrVWVq2SwrJyKyT0qOK9j9lgGmTX6urS0Fah5emTFfOo5r/MOGatqnBVLtm5rLdir4CsMCB8TgVXDkVq6/Pcgs178laVqVZcXeYp8isq2djzMQV3A9Vc49ctda0X6drKT/i48k1nkA5dKpz5QACsVtyXPqKqTmZeE+IdIy4vTy1rYulI8xlOy8rc49T3gg/vHx93lKjImULdrz5ttfLR5qoaqK181ivMqtc3MGUIH3yQAGOAgYNAi+NaFLa9P/ZqfMBatBdWG5kO48tq+YlGOdgdjtys2Jh0Wh5eIIzo9nNWzK6EEgrsXb1LyoM7lmUgHfGdtjhunOnsXVV6kV0DIaxVcq55j+4qrtwdRgBclgFUncjAJjfEnHW1DOtamqhjkpzu7vuSPNsclnAySEzyr2GdznF1h1vFa6FK0Gu7Uc3/AMjkBFQ3wlBK+sA9HbpsEGMNNvwJ4x1PDiBWeejPqocnk+Sv4G+Rt7gyteRLr4S8EWallLqQp3CjZiPc/hX5P/aaR3Hwt4lp19Xm05GDh0YYZGxnB7H6gkSZkV4d4Kmkq8tMfIAwB9Pz695KyBERAREQET4TIXiXH1XK14ZvxfdH/eS3Fk1KavVpWvM5wO3ufoJWeI8Ye3IHoT2B9RHyR/QfrIu/Us7FmYsT7zwLZi9Nzlkx7D6fE84xPYYTOMGc/LbzQwPv8z29Gdx1jy8fPtM1bEbSo1VyMjH57/pPvkBgGx/Kbr080wits7fn2gU/ivhta82IheoDHIuBdSCck0PtgA78n2fpliaRxvjvOTXShqrAKksc3sCACHsO4XCqOQbbb5narayNx+kq/irwOt6NflaXAJBO3Ofbl/17TrzWLHMatIEK+YPUVDhDkeht1Yjqcjf2wRuc7bBVM9FOcgjA3ycnoNj9MfltM2m11diLp9TsqZFVyYDVn+hHz0PRtiHr+JwoAGy+5RptwllW73YOCoxny2G2zYOGzgAHG2UfqtMAvmIc183IT+F8Z5SQMZwcj+e2xjhvFLtO3NTa1ZOM8rHBx05h0PU9feZuKcSFqpUiCuivPIg+e7H3/wC5OSSzNk8PcWTTM7Pp0vDDAD49PyMqRv327Sot3DuM16/HO71atSDWqeXykjfFfOPUMgDkdiQGODgTPxHiprbl1KAWu4rPK9R5qzv5di4YOpJIwwwu5BVs45tY4JJwACScDoM9hnsJb+D+Lg6tVrApyoCajylexSpyC23q3webGcjf7RIlivniTwrVUzWLZ5NWQCHWyxAxAPpeoOQMkrh8EFSMv1mlZr6lrAtvGsKkFKgjpXlFKq11jotjqodgEHudwNjh4lxt83Ili2JYApsaoKxUblUH3Uz8AnAO0gwIm59l/hscR4hZe/mWNk4AAwAqqOioo2VR7DaeNPpHsYKi8zHsJIcC4DZqWHKCEzuxH9Pc9p2jwl4HroUM6kZ3IJ9Tf5vYfwj9NxCKp4H/AGd5xZaAfk/ZHwB95v6f161odElS8qDA7nufrM6qAAAMAbADoJ9kCIiAiIgIiIFd41Y9oYVk+gkPX32P2v4hK6ZL8V4jXVqhWX5LW5mrP4sYJHsT6ht3H0nu7SreCVAW7qV+6/yvsfj/AJnPrmukqF5Z4YTMUxsRuO0+4mGnmt8jEy1HG3+//cxckyKexjDW5Wfyi0e3T2nlekyjO2JUKbPczeqpLdB/2mNtKqgPYeVeoH3m/lI/VcVZwyj93UNsA7n/ADH/AH/OakTUlZqa6zyrh7Pf7qn/AFMguKWecXyxIQZYnu2MhB8DZj9V+Zo8R4ryqGA2GF5jt1ONveaH/UeXSrvuycxPuzjJP6zcjKiaHgJyLNUr16fGcryc+59OQxygbchipzjYHORn494Z1FCZVmt0o/eAg/Y5tiXrz6TjGWAxgjOOgttGproR0scajQ8vKbGrD3acnC4IYAvTliBtkZGw2J8K3kWNetyaqmyklU8ph6d1JRV9P2eYMOo5j7GXajmeJm0OisvcV1rzMcnsAAOrMTsqjuTtLZxfwmr8tmkZQXXn/s7FgwJGeWssN9uzY32BOQJ78NcdyK9Ppq/LYofNxyFrGUfaTnP7xzkkVkdsDaa0eW4FpdNTy2A6rVW+lK6S2xB3K9CoAxlmVsk4wADzV7jHArdL5fmgBnBOMjKkY5lYdiMjfcdRnIOLZreO1aBfK0rvZZgh/M8sqrdyxUZLbZ5FPKCdzkYNQVb9VYWybHOOZj29v5ewAx7CRGiB07k9P/UuXhPwRZewNinsQg9vdvYf767GzeCf2e9LbP8A7HB+vKP0z/xOpaLRpUvKgwO/uT7k9zGiP4HwCvTgYALD2GAv+Udvr1+g2kxESBERAREQEREBNbiGvqora251SterMcD6fJ+BuZsyn/tK8OvrNOprHM9RLBe5BG+B3Ow2+uN8RRTfG3l8XpGood6xW7crFfUeUAZAzlegYd9h9Jr+EvGy8yaXVPy27Cu1iFLntzY+y2e/f+tdsZ6+GamsOyMt9ZBVihw5VCMgj2O3xOY6mhkYhweb57/Oe8q6/WtqLfs+Ft+6/QN8H2MibdOyEqwIM5X+z/8AaR5fLptaxNewS47lfYP7r/F27+47RVqEtULYcjHos64Hye6/Mx1z7alRfJPPLNvU6Vqzg/y9iPcTU1urrpVWsOOY4RQOZ7G/CiDdj+g7kTDTNQSdh/se/wBJ90XiXRZ5a9RU1ucZZsLt+DOBZ7ZXIzIv/p1uq/8AkDy9OcY0ynd9+t9g+129A9PXPNJl6kC8vKvKNguBgD2xjE3OWbWHjWvVVJJ53boBuSO59gPyG8rws2y+MZzy9v5/iP6f1m7rmqqVgiIhPXlAGfriU/i3FuoU4+k1ia2PEnEwa2AO4/0P/qQNWv56qckheStc4JHMVGB+hkbq7SQ2MliJ54PrM6damwUQhmZ2Iqqbl5eY43LkdEGScbDuCNfW69nLDday3MKwzFAcYzgnr8/JxjpJLwt4jfR2EjL0tkWVcxUHIxzDHRhtv36fTLqOH1Xut1S8tBYV8gyGsdF5rGVRkUgjcKzfT2GHjKac1o6AV35KtSnMQAuRzNzHKNkDbJzn4y2kbmr4vWGe/TG1ML5QV7gbVB3HKuP7rquFJPUnGZX7Lec8zBQT15RgE++BsPoJjrQk4AyfYS8+EPAtlzBrFJ78vYD3Y9h+vwZPAg+B+G7dRhsFa9t8bt7BQOv0nZfC3gmulVLoBj7mx393Pf6fn7Ce4NwSvTgYALgY5sdB7KOw/UyUkHwDE+xEBERAREQEREBERAREQKZ468Ii+u16a1Nr4NidrOXcbdObqPn+QnA/EXDwilCDnB5VYepGX7vyP17T9Xym+OPA1WtHmp6L13DDv9RkBvofzB3llTH5Tl28B+P7NERTdl9Kdvdq/lfdf4fy+YPjXBbq9RYllbIQTswKnHTIDYODjIMibqSpwf5fMqv1EnGTZpwdOa7VfHlsxJVfcjG5x+HI+onjRcLCMbXY2XtsbG64/CuNkQfhGBKX+yS3PDgPw2uP1B//AKl81N5A6E/ABMzk1pkttAErPHPEK1ggHeR3iTit4G1bqvuysB+ZlI1OoJO537mVEhr+LtY25OJE23kxp6nsPKoyT+g7knsPmTOg4utY8kWMleD5lmnAD2sT9nzWIKooAGVBzluuxktFc1QCgG0nlYZWpD+8sB9z/hV/xHc74Headjl+XmChV+xWgxWgPsO7HuxyT3M3dcVLOyIqKSSEXoo7AfQYGZrV9JRn4fxK7TsWpsatmHKSpxkH/k79u0ycP0dt7EICSTlidx9Sffr8mSPh/wANWah1yCFPQY3b4A/1nbPDHg1KFU2AZHSsY5R/mx9o/p9Y1FY8E+AcAWPsCN2xuQeyA9B8/wBe3TtJpUqUIihVHb/UnufmZokCIiAiIgIiICIiAiIgIiICIiAiIgV7xh4Sp4hVy2em1c+XaB6kJ7fxKe6/0ODOB8a8LXaS41X15Y8wXl3SxdjlCRuehx1n6dkV4g4HXq6yjjDY9LY3B7H+Uujjn7LAK6L6wdhbzD3GVUYP/wBZK6njFzAhW5RkgkfBmajhb6TUtW6gF1yW7OVIw3ycE7/nuJVuI8ZsotdAFYc5wrLnqfjec5b1NsaiwaHiNhby3c2K2xDZIAMpfEKVNthDYpDlQ3XJB6IPvH4/MgST/wCp3uy6dUQX2EIKUHKcsf8AFsyWQdyikHHXl7zPDvDumfJ8u7VsPS1iFq6hjtUlanCDtk79cbzUhapuqvJBrA5Kwd1zlmI72N33+6PSPk7zUZpY/EXBqkra/TlwqOK7arMF62YEqQwA5kOCNwCCJAcN4dZqH5UG3dj0A+TNI1uUuQqglj0A3MvXg/wK9rAuMnuD9lf8x7n4H/Fp8F+BAqq7DlU7lj/eWfTP2U/0/OdI09C1qFQBVHQCRGjwXglemX0jL43fABPwPYfH55knEQEREBERAREQEREBERAREQEREBERAREQEREDV4joEuXlcdNwe6n3E49498JajSu+orbmqfbzAuGqJ6g/hz2sHvjbv2ueXQMCCAQRgg7gg9iIH5h4RqTprqrgoLVsG5T0OOo/mM7/ADOy+H/HvDK6ErV/JVFAFZRgR+QIY/IJz1kF45/Z1y812mBNfU1jJav5QfeT+HqO23SE8H+BrLyHOMfjxlF/8j8f06yWqmPE+uHFbBVpqmCNyl7CuHt5ObkUA78o52OWx26AZNy8LeDqtMilgC435eoB9/k/7+ZL8F4LVplwgyxHqc45m+p9viSUSfktIiJUIiICIiAiIgIiICIiAiIgIiICIiAiIgIiICIiAiIgJ5RABgAAew2nqICIiAiIgIiICIiAiIgIiICIiAiIgIiICIiAiIgIiICIiAiIgIiICIiAiIgIiICIiAiIgIiICIiAiIgIiICIiAiIgIiICIiAiIgIiICIiAiIgIiICIiAiIgIiICIiAiIgIiICIiB/9k=

**Comment by mrengineer7777** [comment]
@MahdiBabaie Your image didn't paste correctly.  Perhaps just post a web URL?
The PHY chip (LAN8720) and MAC (ESP32) must share a common 50Mhz clock.  Since your module generates 50Mhz, you must feed it into the ESP's GPIO pin 0.
Beware of fake modules.  We just got burned.  Bought 2 LAN8720 modules on Amazon that are apparently knockoffs of Waveshare.  They put out a 50Mhz sine wave instead of clock signal.  Our 8720 modules also don't breakout the PHY enable pin, so it's impossible to disable the clock signal.  ESP32 may not boot properly if there is a clock signal on GPIO 0 during power on.

**Reply by MahdiBabaie** [reply]
Thanks for your response
My Lan8720 like this:
https://imgaz1.staticbg.com/thumb/large/oaupload/banggood/images/3B/23/c581bab4-8def-4a18-94de-8ba4dc0710a9.JPG
Do you mean I have to connect nINT/RETCK of my LAN8720 Module on GPIO0?
Can I use another input instead of GPIO0? For example GPIO35?

**Reply by mrengineer7777** [reply]
Sad news.  That looks exactly like the module that failed for us. Let me know if you find a way to make it work.

>Do you mean I have to connect nINT/RETCK of my LAN8720 Module on GPIO0?

Yes

>Can I use another input instead of GPIO0? For example GPIO35?

For clock input I believe it has to be GPIO0.  https://docs.espressif.com/projects/esp-idf/en/latest/esp32/hw-reference/esp32/get-started-ethernet-kit.html

**Reply by MahdiBabaie** [reply]
As I said before, my LAN8720 Works correctly and can connect to internet and this example:
https://github.com/espressif/arduino-esp32/blob/master/libraries/Ethernet/examples/ETH_LAN8720/ETH_LAN8720.ino
Works correctly on my ESP32 and LAN8720 module. I have not any problem to connect to the internet. My problem is connection to Wireguard server through LAN8720.
You think if I connect my GPIO0 to 50MHz clock of LAN8720 module this problem will be fix?

**Reply by mrengineer7777** [reply]
That's the same example code we used.  Glad it's working for you.  Not sure why Wireguard would be different.  You may have a different issue going on.

**Reply by MahdiBabaie** [reply]
I have only connect my GPIO0 to 50MHz clock? Do I need Anything else? like resistor, capacitor, ... ?

**Reply by mrengineer7777** [reply]
No other hardware. Have to configure ETH.begin() to set GPIO0 as clock input, but I believe that's the default.  FYI I'm a fellow developer, not an employee of espressif.

**Reply by MahdiBabaie** [reply]
Thank you
